src/log-store/src/kafka/util/record.rs

```diff
@@ -0,0 +1,669 @@
+                let offset = producer.produce(kafka_record).await.map(Offset).context(
+                    ProduceRecordSnafu {
+                        topic: &self.ns.topic,
+                    },
+                )?;
+                last_offset = Some(offset);
+            }
+        }
+        // Safety: there must be at least one record produced when the entries are guaranteed not empty.
+        Ok(last_offset.unwrap())
+    }
+}
+
+fn check_records(records: &[Record]) -> Result<()> {
+    let len = records.len();
+    ensure!(
+        len > 0,
+        IllegalSequenceSnafu {
+            error: "Empty sequence"
+        }
+    );
+
+    let mut sequence = Vec::with_capacity(len);
+    let mut entry_ids = HashSet::with_capacity(len);
+    let mut namespaces = HashSet::with_capacity(len);
+    let mut checksum_matched = HashSet::with_capacity(len);
+    for record in records {
+        sequence.push(i32::from(record.tp));
+        entry_ids.insert(record.entry_id);
+        namespaces.insert(&record.ns);
+        checksum_matched.insert(record.checksum == crc32fast::hash(&record.data));
+    }
+
+    ensure!(
+        entry_ids.len() == 1,
+        IllegalSequenceSnafu {
+            error: "Non-unique entry ids"
+        }
+    );
+
+    ensure!(
+        namespaces.len() == 1,
+        IllegalSequenceSnafu {
+            error: "Non-unique namespaces"
+        }
+    );
+
+    ensure!(
+        checksum_matched.len() == 1 && checksum_matched.contains(&true),
+        ChecksumMismatchedSnafu
+    );
+
+    if len == 1 {
+        // A sequence with a single record must contain only a Full record.
+        if sequence[0] != i32::from(RecordType::Full) {
+            return IllegalSequenceSnafu {
+                error: "Missing Full record",
+            }
+            .fail();
+        }
+    } else {
+        // A sequence with multiple records must start with a First record and end with a Last record.
+        // In addition, the Middle records must in order.
+        let prefix = (0..(len - 1) as i32).collect::<Vec<_>>();
+        ensure!(
+            sequence[0..len - 1] == prefix
+                && sequence.last().unwrap() == &i32::from(RecordType::Last),
+            IllegalSequenceSnafu {
+                error: "Out of order",
+            }
+        );
+    }
+
+    Ok(())
+}
+
+fn record_type(seq: usize, num_records: usize) -> RecordType {
+    if seq == 0 {
+        RecordType::First
+    } else if seq == num_records - 1 {
+        RecordType::Last
+    } else {
+        RecordType::Middle(seq)
+    }
+}
+
+fn build_records(entry: EntryImpl, max_record_size: usize) -> Vec<Record> {
+    if entry.data.len() <= max_record_size {
+        let record = Record {
+            version: VERSION,
+            tp: RecordType::Full,
+            checksum: crc32fast::hash(&entry.data),
+            data: entry.data,
+            entry_id: entry.id,
+            ns: entry.ns,
+        };
+        return vec![record];
```

Comment:
Maybe we need to calculate the whole record size cc @v0y4g3r

Reply:
Now an estimated size of a serialized RecordMeta is taken into account. It's simply hardcoded as 256 bytes which is conservative from tests.